tiledb/group.py

```diff
@@ -0,0 +1,204 @@
+
+if TYPE_CHECKING:
+    from .object import Object
+
+
+class Group(lt.Group):
+    """
+    Support for organizing multiple arrays in arbitrary directory hierarchies.
+    Group members may be any number of nested groups and arrays. Members are
+    stored as `tiledb.Object`s which indiciate the member's URI and type.
+
+    Groups may containing associated metadata similar to array metadata where
+    keys are strings. Singleton values may be of type int, float, str, or bytes.
+    Multiple heterogenous values may be placed in containers of type list, tuple,
+    or 1-D np.ndarray. The values within containers are limited to type int or
+    float.
+
+    See more at:
+    https://docs.tiledb.com/main/background/key-concepts-and-data-format#arrays-and-groups
+
+
+    :param uri: The URI to the group
+    :type str
+    :param mode: Read mode ('r') or write mode ('w')
+    :type str
+    :param ctx: A TileDB context
+    :type tiledb.Ctx
+
+    **Example:**
+
+    >>> # Create a group
+    >>> grp_path = "root_group"
+    >>> tiledb.Group.create(grp_path)
+    >>> grp = tiledb.Group(grp_path, "w")
+    >>>
+    >>> # Create an array an add as a member to the group
+    >>> array_path = "array.tdb"
+    >>> domain = tiledb.Domain(tiledb.Dim(domain=(1, 8), tile=2))
+    >>> a1 = tiledb.Attr("val", dtype="f8")
+    >>> schema = tiledb.ArraySchema(domain=domain, attrs=(a1,))
+    >>> tiledb.Array.create(array_path, schema)
+    >>> grp.add(array_path)
+    >>>
+    >>> # Create a group and add is as subgroup
+    >>> subgrp_path = "sub_group"
+    >>> tiledb.Group.create(subgrp_path)
+    >>> grp.add(subgrp_path)
+    >>>
+    >>> # Add metadata to the subgroup
+    >>> grp.meta["ints"] = [1, 2, 3]
+    >>> grp.meta["str"] = "string_metadata"
+    >>> grp.close()
+    >>>
+    >>> # Get the first member's uri and type
+    >>> grp.open("r")
+    >>> member_uri, member_type = grp[0].uri, grp[0].type
+    >>> grp.close()
+    >>>
+    >>> # Remove the subgroup
+    >>> grp.open("w")
+    >>> grp.remove(subgrp_path)
+    >>> grp.close()
+    """
+
+    _NP_DATA_PREFIX = "__np_flat_"
+
+    _mode_to_query_type = {
+        "r": lt.QueryType.READ,
+        "w": lt.QueryType.WRITE,
+    }
+
+    _query_type_to_mode = {
+        lt.QueryType.READ: "r",
+        lt.QueryType.WRITE: "w",
+    }
+
+    class GroupMetadata:
+        """
+        Holds metadata for the associated Group in a dictionary-like structure.
+        """
+
+        GroupMetadataValueType = Union[int, float, str, bytes, np.ndarray]
+
+        def __init__(self, group: "Group"):
+            self._group = group
+
+        def __setitem__(self, key: str, value: GroupMetadataValueType):
+            if not isinstance(key, str):
+                raise TypeError(f"Unexpected key type '{type(key)}': expected str")
+
+            # ensure previous key(s) are deleted (e.g. in case of replacing a
+            # non-numpy value with a numpy value or vice versa)
+            del self[key]
```

Comment:
It does not. The array metadata does not raise an error when trying to delete a key that does not exist either. 

```
>>> with tiledb.open("writing_sparse_multiple", "w") as A:
...     del A.meta["dne"]
...
[no error]
>>>
```

This has to do with the metadata implementation in core where the Array or Group must be open to do any sort of operations on it.
```
def __setitem__(self, key: str, value: GroupMetadataValueType):
...
     if key in self: # It will complain that the Group is not open. But we can't open the Array/Group in here.
          del self[key]
...
```